src/gtk/toga_gtk/widgets/detailedlist.py

```diff
@@ -1,92 +1,231 @@
+        self.right_click_gesture.set_button(3)
+        self.right_click_gesture.set_propagation_phase(Gtk.PropagationPhase.BUBBLE)
+        self.right_click_gesture.connect("pressed", self.gtk_on_right_click)
+
+    def row_factory(self, item):
+        """
+        Args:
+            item (:obj:`Row`)
+        Returns:
+            Returns a (:obj:`TextIconRow`)
+        """
+        new_item = TextIconRow(self.interface.factory, self, item)
+        item._impl = new_item
+        return new_item
+
+    def destroy(self):
+        self.disconnect(self.gtk_on_select_signal_handler)
+        super().destroy()
 
-        self.treeview.set_model(self.store)
+    def change_source(self, source):
+        """
+        Args:
+            source (:obj:`ListSource`)
+        """
+        self.store.remove_all()
+        for item in source:
+            self.store.append(
+                self.row_factory(item))
+
+        # We can't know the dimensions of each row (and thus of the list) until gtk allocates
+        # space for it. Gtk does emit `size-allocate` after allocation but I couldn't find any
+        # guarantees that the rows have their sizes allocated in the order they are inserted
+        # in the `ListStore` and in my opinion that's unlikely to be the case.
+
+        # Therefore we would need to wait for `size-allocate` on all rows and either update
+        # the visibility of the buttons on all `size-allocates` or have a counter and only do
+        # it on the last `size-allocate`. Obviously none of those options are desirable.
+
+        # Fortunately functions added with `idle_add` are run when gtk is idle and thus after
+        # any size allocation. This solves our problem and from the perspective of the user
+        # happens immediately.
+
+        # Even though we are adding the callback to the global loop, it only runs once.
+        # This is what the lambda is for. If a callback returns `False` then it's not ran again.
+        # I used a lambda because returning `False` from `self._list_items_changed()` would mean
+        # returning `False` on success.
+        GLib.idle_add(lambda: not self._list_items_changed())
 
     def insert(self, index, item):
-        self.store.insert(item)
+        """
+        Args:
+            index (int)
+            item (:obj:`Row`)
+        """
+        new_item = self.row_factory(item)
+
+        if index == -1:
+            self.store.append(new_item)
+        else:
+            self.store.insert(index, new_item)
+
+        self.list_box.show_all()
+        self._list_items_changed()
 
     def change(self, item):
-        self.store.change(item)
+        """
+        Args:
+            item (:obj:`Row`)
+        """
+        new_item = self.row_factory(item)
+        index = item._impl.get_index()
+
+        if index == -1:
+            self.store.append(new_item)
+        else:
+            self.store.insert(index, new_item)
```

Comment:
Won't this result in every row change adding a new row to the list?

Reply:
Yes, this must have happened during the above confusion. I added this case to `test_change()`.